server/schedule/operator/operator.go

```diff
@@ -658,40 +658,127 @@ func CreateTransferLeaderOperator(desc string, region *core.RegionInfo, sourceSt
 func CreateMoveRegionOperator(desc string, cluster Cluster, region *core.RegionInfo, kind OpKind, storeIDs map[uint64]struct{}) (*Operator, error) {
-	var steps []OpStep
+	mvkind, steps, err := moveRegionSteps(cluster, region, storeIDs)
+	if err != nil {
+		return nil, err
+	}
+	kind |= mvkind
+	return NewOperator(desc, region.GetID(), region.GetRegionEpoch(), kind, steps...), nil
+}
+
+// moveRegionSteps returns steps to move a region to specific stores.
+//
+// The first store in the slice will not have RejectLeader label.
+// If all of the stores have RejectLeader label, it returns an error.
+func moveRegionSteps(cluster Cluster, region *core.RegionInfo, stores map[uint64]struct{}) (OpKind, []OpStep, error) {
+	storeIDs := make([]uint64, 0, len(stores))
+	for id := range stores {
+		storeIDs = append(storeIDs, id)
+	}
+
+	i, _ := findNoLabelProperty(cluster, opt.RejectLeader, storeIDs)
+	if i < 0 {
+		return 0, nil, errors.New("all of the stores have RejectLeader label")
+	}
+
+	storeIDs[0], storeIDs[i] = storeIDs[i], storeIDs[0]
+	return orderedMoveRegionSteps(cluster, region, storeIDs)
+}
+
+// orderedMoveRegionSteps returns steps to move peers of a region to specific stores in order.
+//
+// If the current leader is not in storeIDs, it will be transferred to a follower which
+//  do not need to move if there is one, otherwise the first suitable new added follower.
+// New peers will be added in the same order in storeIDs.
+// NOTE: orderedMoveRegionSteps does NOT check duplicate stores.
+func orderedMoveRegionSteps(cluster Cluster, region *core.RegionInfo, storeIDs []uint64) (OpKind, []OpStep, error) {
+	var kind OpKind
+
+	oldStores := make([]uint64, 0, len(storeIDs))
+	newStores := make([]uint64, 0, len(storeIDs))
+
+	sourceStores := region.GetStoreIds()
+	targetStores := make(map[uint64]struct{}, len(storeIDs))
+
 	// Add missing peers.
-	for id := range storeIDs {
-		if region.GetStorePeer(id) != nil {
+	var addPeerSteps [][]OpStep
+	for _, id := range storeIDs {
+		targetStores[id] = struct{}{}
+		if _, ok := sourceStores[id]; ok {
+			oldStores = append(oldStores, id)
 			continue
 		}
+		newStores = append(newStores, id)
 		peer, err := cluster.AllocPeer(id)
 		if err != nil {
-			return nil, err
-		}
-		if cluster.IsRaftLearnerEnabled() {
-			steps = append(steps,
-				AddLearner{ToStore: id, PeerID: peer.Id},
-				PromoteLearner{ToStore: id, PeerID: peer.Id},
-			)
-		} else {
-			steps = append(steps, AddPeer{ToStore: id, PeerID: peer.Id})
+			log.Debug("peer alloc failed", zap.Error(err))
+			return kind, nil, err
 		}
+		addPeerSteps = append(addPeerSteps, CreateAddPeerSteps(cluster, id, peer.Id))
+		kind |= OpRegion
 	}
 
+	// Transferring leader to a new added follower may be refused by TiKV.
+	// Ref: https://github.com/tikv/tikv/issues/3819
+	// So, the new leader should be a follower that do not need to move if there is one,
+	// otherwise the first suitable new added follower.
+	storeIDs = append(oldStores, newStores...)
```

Comment:
maybe `candidateLeaderStoreIDs` more better.

Reply:
I don't think `candidateLeaderStoreIDs` is a good name. These store IDs are not just leader candidates. I will use `orderedStoreIDs` to emphasize the order of this slice.